src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Serialization/ObjectWriter.cs

```diff
@@ -278,25 +311,34 @@ public void WriteValue(ReadOnlySpan<byte> span)
                     break;
             }
 
-            var elementType = typeof(byte);
-            Debug.Assert(s_typeMap[elementType] == TypeCode.UInt8);
+            var elementType = typeof(T);
+            Contract.ThrowIfFalse(s_typeMap.ContainsKey(elementType));
+            var actualType = s_typeMap[elementType];
+            Contract.ThrowIfTrue(expectedType != actualType);
 
-            WritePrimitiveType(elementType, TypeCode.UInt8);
+            WritePrimitiveType(elementType, actualType);
+        }
 
-#if NETCOREAPP
-            _writer.Write(span);
-#else
-            // BinaryWriter in .NET Framework does not support ReadOnlySpan<byte>, so we use a temporary buffer to write
-            // arrays of data. The buffer is chosen to be no larger than 8K, which avoids allocations in the large
-            // object heap.
-            var buffer = new byte[Math.Min(length, 8192)];
-            for (var offset = 0; offset < length; offset += buffer.Length)
+        private void WriteSpanPieces<T>(
+            ReadOnlySpan<T> span,
+            int rentLength,
+            Action<BinaryWriter, T[], int> write)
+        {
+            var spanLength = span.Length;
+            var buffer = System.Buffers.ArrayPool<T>.Shared.Rent(Math.Min(spanLength, rentLength));
```

Comment:
>System.Buffers.ArrayPool<T> [](http://example.com/codeflow?start=25&length=27)

I'm curious why you chose to use System.Buffer.ArrayPools instead of ArrayBuilder

Reply:
offline: for straight IO buffers, this feels more in line with what the runtime does.